File: src/main/java/com/study/platform/global/idempotency/IdempotencyResponseWrapper.java
package com.study.platform.global.idempotency;

import jakarta.servlet.ServletOutputStream;
import jakarta.servlet.WriteListener;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.servlet.http.HttpServletResponseWrapper;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;

public class IdempotencyResponseWrapper extends HttpServletResponseWrapper {

    private final ByteArrayOutputStream buffer = new ByteArrayOutputStream();
    private final ServletOutputStream outputStream = new WrappedOutputStream(buffer);
    private final PrintWriter writer = new PrintWriter(outputStream);

    public IdempotencyResponseWrapper(HttpServletResponse response) {
        super(response);
    }

    @Override
    public ServletOutputStream getOutputStream() {
        return outputStream;
    }

    @Override
    public PrintWriter getWriter() {
        return writer;
    }

    @Override
    public void flushBuffer() throws IOException {
        writer.flush();
        outputStream.flush();
        byte[] bytes = buffer.toByteArray();
        getResponse().getOutputStream().write(bytes);
    }

    public String getCapturedBody() {
        writer.flush();
        return buffer.toString(StandardCharsets.UTF_8);
    }

    private static class WrappedOutputStream extends ServletOutputStream {

        private final ByteArrayOutputStream buffer;

        private WrappedOutputStream(ByteArrayOutputStream buffer) {
            this.buffer = buffer;
        }

        @Override
        public void write(int b) {
            buffer.write(b);
        }

        @Override
        public void write(byte[] b) throws IOException {
            buffer.write(b);
        }

        @Override
        public void write(byte[] b, int off, int len) {
            buffer.write(b, off, len);
        }

        @Override
        public boolean isReady() {
            return true;
        }

        @Override
        public void setWriteListener(WriteListener listener) {
        }
    }
}

File: presentation/src/main/kotlin/com/study/platform/global/idempotency/IdempotencyResponseWrapper.kt
package com.study.platform.global.idempotency

import jakarta.servlet.ServletOutputStream
import jakarta.servlet.WriteListener
import jakarta.servlet.http.HttpServletResponse
import jakarta.servlet.http.HttpServletResponseWrapper
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.PrintWriter
import java.nio.charset.StandardCharsets

class IdempotencyResponseWrapper(response: HttpServletResponse) : HttpServletResponseWrapper(response) {

    private val buffer = ByteArrayOutputStream()
    private val outputStream: ServletOutputStream = WrappedOutputStream(buffer)
    private val writer: PrintWriter = PrintWriter(outputStream)

    override fun getOutputStream(): ServletOutputStream = outputStream

    override fun getWriter(): PrintWriter = writer

    @Throws(IOException::class)
    override fun flushBuffer() {
        writer.flush()
        outputStream.flush()
        val bytes = buffer.toByteArray()
        response.outputStream.write(bytes)
    }

    val capturedBody: String
        get() {
            writer.flush()
            return buffer.toString(StandardCharsets.UTF_8)
        }

    private class WrappedOutputStream(private val buffer: ByteArrayOutputStream) : ServletOutputStream() {

        override fun write(b: Int) {
            buffer.write(b)
        }

        @Throws(IOException::class)
        override fun write(b: ByteArray) {
            buffer.write(b)
        }

        override fun write(b: ByteArray, off: Int, len: Int) {
            buffer.write(b, off, len)
        }

        override fun isReady(): Boolean = true

        override fun setWriteListener(listener: WriteListener) {}
    }
}
